### PR TITLE
Remove digest package's dependency on external sha implementation

### PR DESCRIPTION
**Pull request by stevvooe**
The change relies on a refactor of the upstream resumable sha256/sha512 package
that opts to register implementations with the standard library. This allows
the resumable support to be detected where it matters, avoiding unnecessary and
complex code. It also ensures that consumers of the digest package don't need
to depend on the forked sha implementations.

We also get an optimization with this change. If the size of data written to a
digester is the same as the file size, we check to see if the digest has been
verified. This works if the blob is written and committed in a single request.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

### CODE REVIEW

**Comment by stevvooe** [comment]
cc @dmcgowan @jlhawn

**Comment by vbatts** [comment]
so this still does the resumable checksum, but using stdlib crypto?

**Comment by vbatts** [comment]
ah, just moving to another repo and a little refactor, but still a fork of upstream crypto. And looks to have dropped the `noresumabledigest` build tag... except for in a test? :-)

**Comment by dmcgowan** [comment]
@vbatts the build tags are still there, they are only used now in the storage package (and digest test). This prevents us from needing to either use build tags or import the resumable crypto package when vendoring into Docker.

File: registry/storage/blobwriter_resumable.go [hunk 1]
```diff
@@ -2,8 +2,198 @@
...
 
-func (bw *blobWriter) setupResumableDigester() {
-	bw.resumableDigester = digest.NewCanonicalResumableDigester()
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	"github.com/stevvooe/resumable"
+
+	// register resumable hashes with import
+	_ "github.com/stevvooe/resumable/sha256"
```
Comment (dmcgowan) [inline registry/storage/blobwriter_resumable.go]: How does this interact with previous imports of `_ "crypto/sha256"`

Comment (stevvooe) [inline registry/storage/blobwriter_resumable.go]: Presumably, the init function of this package should run after `crypto/sha256`. We can probably force this by making resumable depend on crypto.

**Comment by stevvooe** [comment]
@vbatts This PR minimizes the package that have to depend on resumable digest. It does this by ensuring that only the packages that use resumable digest support import the forked crypto.

Effectively, if you don't use resumable digest support, you don't have to bring the package along at all.

This also has a side-effect that the crypto files in the new package are *identical* to those in the Go distribution.

File: digest/digest.go [hunk 2]
```diff
@@ -15,8 +16,12 @@ import (
 const (
 	// DigestTarSumV1EmptyTar is the digest for the empty tar file.
 	DigestTarSumV1EmptyTar = "tarsum.v1+sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+
 	// DigestSha256EmptyTar is the canonical sha256 digest of empty data
 	DigestSha256EmptyTar = "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+
+	CanonicalAlgorithm = "sha256"
```
Comment (dmcgowan) [inline digest/digest.go]:
```
digest/digest.go:23:2: exported const CanonicalAlgorithm should have comment (or a comment on this block) or be unexported
```

Comment (stevvooe) [inline digest/digest.go]: Fixing this correctly.

File: registry/storage/blobwriter.go [hunk 3]
```diff
@@ -189,33 +185,49 @@ func (bw *blobWriter) validateBlob(ctx context.Context, desc distribution.Descri
...
+		return distribution.Descriptor{}, err
 	}
 
 	if fullHash {
-		digester := digest.NewCanonicalDigester()
-
-		digestVerifier, err := digest.NewDigestVerifier(desc.Digest)
-		if err != nil {
-			return distribution.Descriptor{}, err
+		// a fantastic optimization: if the the written data and the size are
```
Comment (dmcgowan) [inline registry/storage/blobwriter.go]: This is great, I assumed we already had this optimization

**Comment by dmcgowan** [comment]
LGTM

File: digest/digester.go [hunk 4]
```diff
@@ -1,54 +1,95 @@
...
+		return false
+	}
+
+	// check availability of the hash, as well
+	return h.Available()
+}
+
+// New returns a new digester for the specified algorithm. If the algorithm
+// does not have a digester implementation, nil will be returned. This can be
+// check prior be calling Available before calling New.
```
Comment (tiborvass) [inline digest/digester.go]: `by calling`

File: digest/digester.go [hunk 5]
```diff
@@ -1,54 +1,95 @@
...
+	return d.hash
 }
 
-// ResumableDigester is a digester that can export its internal state and be
-// restored from saved state.
-type ResumableDigester interface {
-	ResumableHash
-	Digest() Digest
+func (d *digester) Digest() Digest {
+	return NewDigest(d.alg, d.Hash())
```
Comment (tiborvass) [inline digest/digester.go]: `d.hash`

File: digest/digester_resumable_test.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+// TestResumableDetection just ensures that the resumable capability of a hash
+// is exposed through the digester type, which is just a hash plus a Digest
+// method.
+func TestResumableDetection(t *testing.T) {
+	d := Canonical.New()
+
+	if _, ok := d.Hash().(resumable.Hash); !ok {
+		t.Fatalf("expected digester to implement resumable: %#v, %v", d, d.Hash())
```
Comment (tiborvass) [inline digest/digester_resumable_test.go]: `implement resumable.Hash`

File: registry/storage/blobwriter.go [hunk 7]
```diff
@@ -15,14 +12,19 @@ import (
 	storagedriver "github.com/docker/distribution/registry/storage/driver"
 )
 
+var (
+	errResumableDigestNotAvailable = fmt.Errorf("resumable digest not available")
```
Comment (tiborvass) [inline registry/storage/blobwriter.go]: `errors.New`

File: registry/storage/blobwriter_resumable.go [hunk 8]
```diff
@@ -2,8 +2,198 @@
...
+	if offset == int64(h.Len()) {
+		// State of digester is already at the requested offset.
+		return nil
+	}
+
+	// List hash states from storage backend.
+	var hashStateMatch hashStateEntry
+	hashStates, err := bw.getStoredHashStates(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get stored hash states with offset %d: %s", offset, err)
```
Comment (tiborvass) [inline registry/storage/blobwriter_resumable.go]: `%v` for `err`

File: registry/storage/blobwriter_resumable.go [hunk 9]
```diff
@@ -2,8 +2,198 @@
...
+			break // Found an exact offset match.
+		} else if hashState.offset < offset && hashState.offset > hashStateMatch.offset {
+			// This offset is closer to the requested offset.
+			hashStateMatch = hashState
+		} else if hashState.offset > offset {
+			// Remove any stored hash state with offsets higher than this one
+			// as writes to this resumed hasher will make those invalid. This
+			// is probably okay to skip for now since we don't expect anyone to
+			// use the API in this way. For that reason, we don't treat an
+			// an error here as a fatal error, but only log it.
```
Comment (tiborvass) [inline registry/storage/blobwriter_resumable.go]: duplicate `an`

File: registry/storage/blobwriter_resumable.go [hunk 10]
```diff
@@ -2,8 +2,198 @@
...
+	// the requested offset.
+	for _, hashState := range hashStates {
+		if hashState.offset == offset {
+			hashStateMatch = hashState
+			break // Found an exact offset match.
+		} else if hashState.offset < offset && hashState.offset > hashStateMatch.offset {
+			// This offset is closer to the requested offset.
+			hashStateMatch = hashState
+		} else if hashState.offset > offset {
+			// Remove any stored hash state with offsets higher than this one
```
Comment (tiborvass) [inline registry/storage/blobwriter_resumable.go]: clarify `this one`, `hashState.offset` or `offset` ?

File: registry/storage/blobwriter_resumable.go [hunk 11]
```diff
@@ -2,8 +2,198 @@
...
+			// as writes to this resumed hasher will make those invalid. This
```
Comment (tiborvass) [inline registry/storage/blobwriter_resumable.go]: clarify `those` (sorry if it's obvious). Are you referring to `writes` or `offsets` ?

File: registry/storage/blobwriter_resumable.go [hunk 12]
```diff
@@ -2,8 +2,198 @@
...
+	}
+
+	hashStateEntries := make([]hashStateEntry, 0, len(paths))
+
+	for _, p := range paths {
+		pathSuffix := path.Base(p)
+		// The suffix should be the offset.
+		offset, err := strconv.ParseInt(pathSuffix, 0, 64)
+		if err != nil {
+			logrus.Errorf("unable to parse offset from upload state path %q: %s", p, err)
```
Comment (tiborvass) [inline registry/storage/blobwriter_resumable.go]: `%v`

**Comment by tiborvass** [comment]
Apart from those minor mostly irrelevant nits, things look good, but since IANTM, I give you a weak LGTM

**Comment by stevvooe** [comment]
@tiborvass Most of those edits were already present in the moved source.

**Comment by tiborvass** [comment]
@stevvooe true, feel free to ignore my comments :)

**Pull request merged**